CLDSRV-374 putMetadata API route is not updating null version properly

nicolas2bert opened the pull request:
Instead of using the provided "null" value, the metadata "null version id" is used when updating the metadata of a null version.

Note: In S3C, the use of the "putMetadata" call is restricted to versioned buckets.

Issue: CLDSRV-374

nicolas2bert commented:
ping

bert-e commented:
# Hello nicolas2bert,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/cloudserver/doc/user).





*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-374 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `7.10.28`

* `7.70.21`

* `8.5.21`

* `8.6.10`

* `8.7.21`


Please check the `Fix Version/s` of CLDSRV-374, or the target
branch of this pull request.

nicolas2bert commented:
ping

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/8.5/bugfix/CLDSRV-374/put-metadata-null` with contents from `w/7.70/bugfix/CLDSRV-374/put-metadata-null`
and `development/8.5`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/8.5/bugfix/CLDSRV-374/put-metadata-null origin/development/8.5
 $ git merge origin/w/7.70/bugfix/CLDSRV-374/put-metadata-null
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.5/bugfix/CLDSRV-374/put-metadata-null
```

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/CLDSRV-374/put-metadata-null` into
`development/7.10`
* [w/7.70/bugfix/CLDSRV-374/put-metadata-null](https://github.com/scality/cloudserver/compare/development/7.70...w/7.70/bugfix/CLDSRV-374/put-metadata-null) will be merged into `development/7.70` 
* [w/8.5/bugfix/CLDSRV-374/put-metadata-null](https://github.com/scality/cloudserver/compare/development/8.5...w/8.5/bugfix/CLDSRV-374/put-metadata-null) will be merged into `development/8.5` 
* [w/8.6/bugfix/CLDSRV-374/put-metadata-null](https://github.com/scality/cloudserver/compare/development/8.6...w/8.6/bugfix/CLDSRV-374/put-metadata-null) will be merged into `development/8.6` 
* [w/8.7/bugfix/CLDSRV-374/put-metadata-null](https://github.com/scality/cloudserver/compare/development/8.7...w/8.7/bugfix/CLDSRV-374/put-metadata-null) will be merged into `development/8.7` 



The following branches will **NOT** be impacted:

* `development/7.4`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

jonathan-gramain commented on lib/routes/routeBackbeat.js:
```diff
@@ -472,14 +472,29 @@ function putMetadata(request, response, bucketInfo, objMd, log, callback) {
         // versionId
         const options = {
-            versioning: bucketInfo.isVersioningEnabled(),
+            versioning,
```
It could be cleaner to
1. prepare the final `versioning` and `versionId` fields to pass in options
2. build the `options` object with those two fields

nicolas2bert replied:
I did not change the existing logic much to have the change easily forwarded into development/8.x branches which have a slightly different logic.

jonathan-gramain commented on lib/routes/routeBackbeat.js:
```diff
@@ -472,14 +472,29 @@ function putMetadata(request, response, bucketInfo, objMd, log, callback) {
+            if (versioning && !nullVersionId) {
+                // If the null version does not have a version id, it is a current null version.
+                // To update the metadata of a current version, versioning is set to false.
+                options.versioning = false;
```
Beware of this, as option fields are converted to strings when passed to Metadata (through the query string), there's a good chance that we interpret `versioning: "false"` as being true in Metadata, I'd rather recommend to not pass `versioning` at all if it is false. Metadata should have better parsing of its options on the other hand but I don't think it's done properly now.

bert-e commented:
# History mismatch



Merge commit #048a10b191caa788ca526a2690d2a75177e0ad1a on the integration branch
`w/7.70/bugfix/CLDSRV-374/put-metadata-null` is merging a branch which is neither the current
branch `bugfix/CLDSRV-374/put-metadata-null` nor the development branch
`development/7.70`.

It is likely due to a rebase of the branch `bugfix/CLDSRV-374/put-metadata-null` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

jonathan-gramain commented on lib/routes/routeBackbeat.js:
```diff
@@ -472,22 +472,49 @@ function putMetadata(request, response, bucketInfo, objMd, log, callback) {
             omVal.replicationInfo.isNFS = !isReplica;
         }
+
+        // NOTE: When 'versioning' is set to true and no 'versionId' is specified,
+        // it results in the creation of a "new" version, which also updates the master.
```
Possibly this comment could fit better above `if (versioning) {`

jonathan-gramain commented on tests/functional/raw-node/test/routes/routeBackbeat.js:
```diff
@@ -172,6 +189,155 @@ describeSkipIfAWS('backbeat routes', () => {
+                    .then(() => done());
+            });
+
+            it('should update metadata of a current null version', done => {
```
Would you mind adding some similar tests on versioning-suspended buckets:

- when there is an existing null version
- when there is no existing null version

and make sure we only delete the data when a prior null version exists

nicolas2bert replied:
In S3C, the use of the "putMetadata" call is restricted to versioned buckets.
I will add tests for Zenko/Artesca branches that support PUT metadata for non-versioned buckets.

jonathan-gramain commented on lib/routes/routeBackbeat.js:
```diff
@@ -502,7 +529,10 @@ function putMetadata(request, response, bucketInfo, objMd, log, callback) {
+                    // - if a version ID is provided in the query string (updating MD of a specific version) OR
+                    // - if no version ID provided and versioning is not enabled (updating MD of the master key),
+                    // data locations are allowed to be removed as they will be overwritten.
+                    (versionId || !versioning) &&
```
I have a doubt that we may remove valid data in the following case, not completely sure, to be checked:

- versioning-suspended bucket
- no existing null version

I am thinking that if `objMd` is the metadata of the current master (a real version), we may remove the master's data and create a new null version, making the old master unreadable.

nicolas2bert replied:
If there is no existing null version, the object MD (objMd) of the null version will be undefined, resulting in the condition being false.

Kerkesni commented on lib/routes/routeBackbeat.js:
```diff
@@ -472,22 +472,61 @@ function putMetadata(request, response, bucketInfo, objMd, log, callback) {
+                    // standalone master keys.
+                    versioning = false;
+                } else {
+                    const versioningConf = bucketInfo.getVersioningConfiguration();
+                    // The purpose of this condition is to address situations in which
+                    // - versioning is "suspended" and
+                    // - no existing object or no null version.
+                    // In such scenarios, we generate a new null version and designate it as the master version.
+                    if (versioningConf && versioningConf.Status === 'Suspended') {
+                        versionId = '';
+                        omVal.isNull = true;
+                    }
```
There might be a case we are missing here (maybe only applicable for the Zenko branches) :
- Create a bucket with versioning suspended
- Put an object
- Enable Versioning
- Put a second version of the same object
- Suspend versioning
- Call PutMetadata on the null version

This will cause Cloudserver to only update the master object without the versions which might break things

welansari commented on lib/routes/routeBackbeat.js:
```diff
@@ -472,22 +472,61 @@ function putMetadata(request, response, bucketInfo, objMd, log, callback) {
+        let versionId = decodeVersionId(request.query);
+        let versioning = bucketInfo.isVersioningEnabled();
+
+        if (versionId === 'null') {
+            // Retrieve the null version id from the object metadata.
+            versionId = objMd && objMd.versionId;
+            if (!versionId) {
+                if (versioning) {
+                    // If the null version does not have a version id, it is a current null version.
+                    // To update the metadata of a current version, versioning is set to false.
+
+                    // This condition is to handle the case where a single null version looks like a master
+                    // key and will not have a duplicate versioned key and no version ID.
+                    // They are created when you have a non-versioned bucket with objects,
+                    // and then convert bucket to versioned.
+                    // If no new versioned objects are added for given object(s), they look like
+                    // standalone master keys.
+                    versioning = false;
+                } else {
+                    const versioningConf = bucketInfo.getVersioningConfiguration();
+                    // The purpose of this condition is to address situations in which
+                    // - versioning is "suspended" and
+                    // - no existing object or no null version.
+                    // In such scenarios, we generate a new null version and designate it as the master version.
+                    if (versioningConf && versioningConf.Status === 'Suspended') {
+                        versionId = '';
+                        omVal.isNull = true;
+                    }
+                }
+            }
+        }
```
My comment might only be applicable for Zenko, but this doesn't work in versioning suspended mode.
When having a null version created after suspending the versioning, the options will equal `{ versionId: '$VERSION_ID' }` which will create another version instead of updating the master object.

nicolas2bert replied:
I'm unsure about which of the tests I wrote is incorrect. Please feel free to identify the test that is wrong.

nicolas2bert replied:
I just ran the tests from https://github.com/scality/cloudserver/pull/5171 and they all passed.

welansari replied:
Yes the tests pass as what i pointed out is not visible from the S3 API side.
Basically the object initially has only a master document in mongo and instead of only updating that, we create a new version object as well as update the mater.
I'm not entirely sure if this is an intended behaviour or not hence why i pointed it out.

nicolas2bert committed:
CLDSRV-374 putMetadata API route is not updating null version properly

Instead of using the provided "null" value, the metadata "null version id" is now used when updating the metadata of a null version.

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-374 contains:


* `7.10.28`

* `7.70.21`

* `8.5.21`

* `8.6.10`

* `8.7.21`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.7.21`


Please check the `Fix Version/s` of CLDSRV-374, or the target
branch of this pull request.

nicolas2bert commented:
While it's not the preferred scenario, due to the urgent need to implement this fix in Artesca, we will initially integrate it into the development/8.7 branch and subsequently backport it to the development/7.x branches.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

jonathan-gramain commented on tests/multipleBackend/routes/routeBackbeat.js:
```diff
@@ -274,6 +303,982 @@ describe.skip('backbeat routes', () => {
             .then(() => done());
     });
 
+    describe.only('null version', () => {
```
`.only`

nicolas2bert replied:
We eliminated the '.skip' modifier in the 'describe' section of the parent. 
Initially, all tests were being skipped. 
However, I unskipped them and only executed the null test for the time being. 

To address the remaining routeBackbeat tests, I created a follow-up task CLDSRV-394, which is mentioned here: https://github.com/scality/cloudserver/pull/5167/files#diff-e346bc0177b28612b0892a923d8c1d33429b4d7a3c8bd955aadf6e3ccb701dcdR1281.

jonathan-gramain replied:
It may cause confusion to have a committed `.only` test, usually we follow the pattern of skipping tests until they pass, I think we should follow this scheme here too, so skipping the rest of the tests in this suite and removing `.only` of the only test we'd like to activate may be more appropriate.

nicolas2bert replied:
I agree.

jonathan-gramain commented on tests/multipleBackend/routes/routeBackbeat.js:
```diff
@@ -274,6 +303,982 @@ describe.skip('backbeat routes', () => {
+        it('should create a new null version if versioning suspended and no version', done => {
+            let objMD;
+            return async.series([
+                next => s3.putBucketVersioning({ Bucket: bucket, VersioningConfiguration: { Status: 'Enabled' } },
```
You don't need to enable versioning before suspending it, you can just put a versioning-suspended status on a nonversioned bucket (same for the other tests).

nicolas2bert replied:
Indeed, I simply prefer having it Enabled prior to Suspending it as it feels more instinctive to me. 
However, as enabling it beforehand is not obligatory, I will eliminate that step.

jonathan-gramain commented on tests/multipleBackend/routes/routeBackbeat.js:
```diff
@@ -274,6 +303,982 @@ describe.skip('backbeat routes', () => {
+                    next),
+                next => s3.putBucketVersioning({ Bucket: bucket, VersioningConfiguration: { Status: 'Suspended' } },
+                    next),
+                next => s3.putObject({ Bucket: bucket, Key: keyName, Body: new Buffer(testData) }, next),
```
Is it useful to PUT and DELETE an object prior to using the backbeat PUT?

nicolas2bert replied:
For the test to be closer to reality and to be able "GET" the null version metadata that I will later "PUT".

jonathan-gramain commented on tests/multipleBackend/routes/routeBackbeat.js:
```diff
@@ -274,6 +303,982 @@ describe.skip('backbeat routes', () => {
+                    next),
+                next => s3.putBucketVersioning({ Bucket: bucket, VersioningConfiguration: { Status: 'Suspended' } },
+                    next),
+                next => s3.putObject({ Bucket: bucket, Key: keyName, Body: new Buffer(testData) }, next),
```
There's probably no need to PUT an object first before DELETE, as a DELETE would normally put a delete marker with a null version.

nicolas2bert replied:
Same reason as earlier. 
Also, we create the object to GET its metadata. 
In my opinion, this appears to be a more realistic scenario.

jonathan-gramain commented:
Thanks for the extensive test coverage @nicolas2bert

welansari commented on lib/routes/routeBackbeat.js:
```diff
@@ -491,24 +491,67 @@ function putMetadata(request, response, bucketInfo, objMd, log, callback) {
+        // NOTE: Since option fields are converted to strings when they're sent to Metadata via the query string,
+        // Metadata interprets the value "false" as if it were true.
+        // Therefore, to avoid this confusion, we don't pass the versioning parameter at all if its value is false.
+        if (versioning) {
+            options.versioning = true;
+        }
+
```
i can see one case where this won't work:
- create non versioned bucket
- put object
- put object versioning : Enabled
- put metadata

Replication uses the versionId in the actual MD when calling this function (on 8.x)
In the example i gave it will set `undefined` as version id which will create a version and update the master, which means that we loose the null (non versioned) version

welansari replied:
This case actually happened with KB, as they use OOB to copy metadata from an S3C to Artesca and then replicate them, the S3C bucket they used was initially non versioned and contained non versioned objects which they then turned into a versioned bucket to work with OOB.

welansari replied:
maybe this case should be handled in backbeat itself, i'm not really sure

nicolas2bert replied:
Is "should update metadata of a current null version" not covering this test?

nicolas2bert replied:
Indeed `versionId` will be `undefined`, but versioning will be `false`. Only the master should be updated.
This case is handled by setting `versioning = false;` explicitly.

welansari replied:
> Is "should update metadata of a current null version" not covering this test?

Not really cause backbeat doesn't set `null` as version id in the query params, it'll put `undefined` in there, and this is what creates the problem

welansari replied:
i'm resolving this comment as this is a backbeat issue and is easier to address in backbeat

nicolas2bert replied:
If the version ID specified in the query parameters is anything other than "null," it will fail the "null" validation. https://github.com/scality/cloudserver/pull/5167/files#diff-6e9c89f5d34a58972fb42340be6bdaff29f9167883a5265dc64c2d8d384ca31eR504

nicolas2bert commented:
@bert-e approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/8.7`



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.5`
* `development/8.6`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/8.7`



The following branches have **NOT** changed:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.5`
* `development/8.6`



Please check the status of the associated issue CLDSRV-374.

Goodbye nicolas2bert.
